### PR TITLE
refactor improve performance:

### PR DESCRIPTION
**Pull request by ronenhamias**
* reduce the preasure on class for name on each request by caching the return type when building a proxy.
* do not call ServiceLoader on each request.
* improve tests not create gateway on each test.

Test included:
```java
package io.scalecube.services.benchmarks;

import io.scalecube.services.Microservices;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

import reactor.core.publisher.Flux;

public class ServicesBenchmark {

  public static AtomicInteger total = new AtomicInteger(0);

  public static void main(String[] args) throws InterruptedException, ExecutionException {
    Microservices gateway = Microservices.builder().build().startAwait();

    Microservices node = Microservices.builder()
        .seeds(gateway.cluster().address())
        .services(new BenchmarkServiceImpl())
        .build()
        .startAwait();

    Flux.from(gateway.call().create().api(BenchmarkService.class)
        .requestMany(new BenchmarkMessage("hello joe", 1_000_000)))
        .blockLast();
    Thread.sleep(1000);

    ExecutorService exec = Executors.newFixedThreadPool(6);
    long both = System.currentTimeMillis();
    CompletableFuture.allOf(
        runClient(gateway, 1_000_000, exec),
        runClient(gateway, 1_000_000, exec),
        runClient(gateway, 1_000_000, exec),
        runClient(gateway, 1_000_000, exec),
        runClient(gateway, 1_000_000, exec),
        runClient(gateway, 1_000_000, exec))
        .get();
    long done = System.currentTimeMillis();
    System.out.println("requested : " + total + " :  at rate of(ms): " + (total.get() / (done - both)));
    exec.shutdownNow();
  }

  static AtomicInteger executerNumber = new AtomicInteger();

  static CompletableFuture<Void> runClient(Microservices gateway, int count, ExecutorService exec) {
    BenchmarkMessage request = new BenchmarkMessage("hello joe", count);
    total.addAndGet(count);
    return CompletableFuture.runAsync(() -> {
      int myExecutionNumber = executerNumber.incrementAndGet();
      System.out.println("### Task no." + myExecutionNumber + " request for(" + count + ")");
      long start = System.currentTimeMillis();
      Flux.from(gateway.call().create().api(BenchmarkService.class)
          .requestMany(request))
          .blockLast();
      long end = System.currentTimeMillis();
      System.out.println("request many " + myExecutionNumber + ": " + (end - start));
    }, exec);
  }
}
```

### CODE REVIEW

**Comment by ronenhamias** [comment]
After this fix i reach 90,000 stream events per sec (locally) for request many where i ask 1M messages for service and block till last.

request reply does 12K and still require deeper investigation.

**Comment by segabriel** [comment]
I  ran benchmarks (https://github.com/scalecube/scalecube-services/pull/109) on my home laptop
results (for test requestOne, nThreads=4, duretion=5min) were following:

    count=834933, 
    min=1.15303, 
	max=801.103885, 
	mean=88.53025134779888, 
	stddev=49.85243414974958, 
	median=84.465437, 
	p75=121.347366, 
	p95=172.00612999999998, 
	p98=195.691731, 
	p99=213.447085, 
	p999=342.67309, 
	mean_rate=2783.1789277083512, 
	m1=2925.5056291887413, 
	m5=1952.4524752561217, 
	m15=1119.674823708388, 
	rate_unit=events/second, duration_unit=milliseconds

and after applying these improvements:

	count=3721783, 
	min=0.124098, 
	max=129.73688099999998, 
	mean=19.549434026790586, 
	stddev=16.914100802323368, 
	median=16.497311, 
	p75=33.275011, 
	p95=49.409980999999995, 
	p98=57.662079, 
	p99=64.08773, 
	p999=77.05458399999999, 
	mean_rate=12406.269157319271, 
	m1=12913.238396023744, 
	m5=8970.266334641301, 
	m15=5549.400468304719, 
	rate_unit=events/second, duration_unit=milliseconds

so it's very nice improvements!

File: services/src/main/java/io/scalecube/services/ServiceCall.java [hunk 1]
```diff
@@ -187,33 +181,30 @@ public ServiceCall create() {
...
 
-      Metrics.mark(serviceInterface, metrics, method, "request");
-      Class<?> parameterizedReturnType = Reflect.parameterizedReturnType(method);
-      boolean isRequestTypeServiceMessage = Reflect.isRequestTypeServiceMessage(method);
-      CommunicationMode mode = Reflect.communicationMode(method);
-
       ServiceMessage request = ServiceMessage.builder()
           .qualifier(Reflect.serviceName(serviceInterface), method.getName())
-          .data(method.getParameterCount() != 0 ? args[0] : NullData.NULL_DATA)
+          .data(method.getParameterCount() != 0 ? params[0] : NullData.NULL_DATA)
```
Comment (aharonha) [inline services/src/main/java/io/scalecube/services/ServiceCall.java:195]: Is it just me or `!=0` may imply more than 1 parameter?

File: services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageDataCodec.java [hunk 2]
```diff
@@ -15,22 +15,21 @@
 import org.slf4j.LoggerFactory;
 
 import java.nio.charset.Charset;
-import java.util.Optional;
 
 public final class ServiceMessageDataCodec {
```
Comment (aharonha) [inline services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageDataCodec.java]: What is the difference between `ServiceMessageDataCodec` and `ServiceMessageCodec`?

File: services/src/main/java/io/scalecube/services/ServiceCall.java (hunk 1 above)
Comment (ronenhamias) [reply services/src/main/java/io/scalecube/services/ServiceCall.java:195]: yes no more then one paramter

File: services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageDataCodec.java (hunk 2 above)
Comment (ronenhamias) [reply services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageDataCodec.java]: service message decode headers and not body
and service data codec decode only data

File: services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageCodec.java [hunk 3]
```diff
@@ -38,8 +36,6 @@ public ServiceMessageCodec(HeadersCodec headersCodec) {
     } else if (message.hasData()) {
       dataBuffer = ByteBufAllocator.DEFAULT.buffer();
       try {
-        String contentType = Optional.ofNullable(message.dataFormat()).orElse(DEFAULT_DATA_FORMAT);
-        DataCodec dataCodec = DataCodec.getInstance(contentType);
```
Comment (artem-v) [inline services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageCodec.java]: Those two lines were  correct. What was not is implementation of the `DataCodec.getInstance()`. It has to cache once loaded codec classes and on subsequent calls return from hash map.

File: services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageDataCodec.java [hunk 4]
```diff
@@ -15,22 +15,21 @@
...
+  public static final DataCodec dataCodec = DataCodec.getInstance(DEFAULT_DATA_FORMAT);
+
   public ServiceMessage encode(ServiceMessage message) {
     if (message.hasData(ByteBuf.class)) {
       return message;
     } else if (message.hasData()) {
       ByteBuf buffer = ByteBufAllocator.DEFAULT.buffer();
       try {
-        String contentType = Optional.ofNullable(message.dataFormat()).orElse(DEFAULT_DATA_FORMAT);
-        DataCodec dataCodec = DataCodec.getInstance(contentType);
```
Comment (artem-v) [inline services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageDataCodec.java]: Those two lines were  correct. What was not is implementation of the `DataCodec.getInstance()`. It has to cache once loaded codec classes and on subsequent calls return from hash map.

File: services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageDataCodec.java [hunk 5]
```diff
@@ -51,8 +50,6 @@ public ServiceMessage decode(ServiceMessage message, Class<?> type) {
     Class<?> targetType = ExceptionProcessor.isError(message) ? ErrorData.class : type;
 
     try (ByteBufInputStream inputStream = new ByteBufInputStream(((ByteBuf) message.data()).slice())) {
-      String contentType = Optional.ofNullable(message.dataFormat()).orElse(DEFAULT_DATA_FORMAT);
-      DataCodec dataCodec = DataCodec.getInstance(contentType);
```
Comment (artem-v) [inline services-codec/src/main/java/io/scalecube/services/codec/ServiceMessageDataCodec.java]: Those two lines were  correct. What was not is implementation of the `DataCodec.getInstance()`. It has to cache once loaded codec classes and on subsequent calls return from hash map.

**Review by artem-v — CHANGES_REQUESTED** [review]

**Review by artem-v — APPROVED** [review]

**Pull request merged**